### PR TITLE
Integrate YellowFin Optimizer(with test) in T2T

### PR DESCRIPTION
**Pull request by ReDeiPirati**
I have adapted [YellowFin](https://github.com/JianGoForIt/YellowFin) to be usable within `T2T` and to be `Tf.train.Optimizer` "compliant", but unfortunately a possible error/TF bug does not ensure the smooth running of the Optimizer. The Optimizer is far to be "production-ready" and there will be different improvements to be done such as multi-GPU testing, code readability and TF integration. This PR is in request of #125 .

### CODE REVIEW

**Comment by lukaszkaiser** [comment]
Thanks!

**Comment by lukaszkaiser** [comment]
Ah, sorry. Your licence will be automatically overwritten in the next CL. This is an Apache licensed project and we have no say in it, if you're not ok with that, I'm afraid you'll need to send a PR removing your code. (I hope it's not a problem because Apache and MIT are so close, but I need to say this.)

**Comment by ReDeiPirati** [comment]
It's not a problem for me. I have forgotten to change the license taken from the official YelloFin repo with the Google Apache license. One last question: the switch of license will override the initial YellowFin author, do you think that i have to add a comment line who say something like: "Code adapted from the official_yellofin_repo_link"?

**Comment by ReDeiPirati** [comment]
Ah another thing: i haven't claimed in the `README.md` about the existence of the new Optimizer, do you thing that it's better to take a "low profile" until we are enough confident to be Experiment ready? Or do you prefer to announce it's existence, maybe adding a label such as experimental feature to warn the community?

**Comment by lukaszkaiser** [comment]
I think it's fine. I'm adding you (and others) to AUTHORS right now, and changing all our licences to say Tensor2Tensor authors. Are you the author of YellowFin? Should I add someone else?

**Comment by ReDeiPirati** [comment]
No i'm not the author of TF-YF it's @JianGoForIt.

**Comment by lukaszkaiser** [comment]
Could you ask him if it's ok to change the licence? I don't think we can do that without the author's approval, sorry!

**Comment by lukaszkaiser** [comment]
But I'll wait with bringing this back -- I think we just need an OK since these licences are so similar.

**Comment by ReDeiPirati** [comment]
No problem, it's the right to do!!

**Comment by JianGoForIt** [comment]
Hi @ReDeiPirati @lukaszkaiser,

Thanks for the info. It would be great to get YF into T2T. Please go ahead to change the license. I will double check with my collaborators and I think they should also be happy as long as things are properly cited.

I am currently adding some automatic clipping feature to YF which can avoid quite some of the issues reported regarding exploding gradient. If you want to wait this feature before merging in, I can keep you updated. Hopefully, it should be done before early next week.

**Comment by JianGoForIt** [comment]
Hi @ReDeiPirati @lukaszkaiser 

We actually switched license to APACHE2.0 now so that you can directly use it without overwriting. 

Cheers

**Comment by ReDeiPirati** [comment]
Thanks to you @JianGoForIt  and your collaborators for the support!!

**Comment by lukaszkaiser** [comment]
That's wonderful, enormous thanks guys!

**Comment by lukaszkaiser** [comment]
@ReDeiPirati : I did the changes and it's part of 1.0.14. But the test was failing for me, I had to set `n_iters=0` to disable it. Could you take a looks and correct it? Did it ever work for you?

**Comment by lukaszkaiser** [comment]
@JianGoForIt : it's wonderful, I'd love to hear about automatic clipping and merge it when it's done. I want to run our Transformer models with YellowFin, hopefully next week. Thanks!

**Comment by ReDeiPirati** [comment]
@lukaszkaiser, Mh that it's strange, the test has never failed during my test. The Optimzer works only sometimes, as i said in the issue in an "undeterministic" way(even without changing a single line of code, but trying multiple times).

**Comment by JianGoForIt** [comment]
@lukaszkaiser @ReDeiPirati 

Thanks for the effort. Typically this kind of none deterministic failure might be related to some exploding gradient, it can be avoided by carefully set gradient clipping threshold.

**Comment by ReDeiPirati** [comment]
@JianGoForIt have you experimented the optimizer with the `tf.Estimator` API? I think that our problem could be related to a bad graph integration with this API, but i'm not completely sure. I'll keep investigating.
Thank you for the advice about the exploding gradient, i have also experimented that failure!

**Comment by JianGoForIt** [comment]
Hi @lukaszkaiser @ReDeiPirati,

Thanks for the efforts in merging YF into Tensor2Tensor repo.

We have recently implemented and tested the following updates:

1. Multiple GPU compatibility.
2. Replaced numpy root with mathematically equivalent closed form solution (using Vieta's substitution) for the cubic root solver. It eliminate stability issues of numpy root function, which might use iterative procedures inside.
3. Debiasing the estimators for sparse gradients (e.g. when word embedding is used)
4. The code framework is refactored with tf.optimizer as base class. And the global step issue is not an problem anymore.

I think these changes might be useful for T2T applications. Please merge them (or maybe just copy the new yellowfin.py to T2T repo :) ) if they are relevant to you.

Thanks in advance.

**Comment by ReDeiPirati** [comment]
Hi @JianGoForIt 
I've followed the last PR of YellowFin, this update seems really great!!!
@lukaszkaiser, i will take care the update this weekend.

**Comment by JianGoForIt** [comment]
Hi, Thanks for following up and helping integrate. Please keep us updated.

Cheers

Ph. D. Candidate in Computer Science,
Stanford University.

Homepage: http://cs.stanford.edu/people/zjian/

2017-08-17 11:04 GMT-07:00 K1ngP1r@t& <notifications@github.com>:
